### PR TITLE
Create `AST` version of `insertBlockE`, prove its contract (after refining it)

### PR DESCRIPTION
**Pull request by mark-moir**
This PR provides the first experience with using the new framework to prove something about LBFT.  Specifically, it adds an `AST` version of `insertBlockE`, and proves its contract (after refining it).  The refinement required more assumptions for some proof obligations.  This required changes to the proofs "up the stack"~~, which have not all been done.  There is a comment where it still breaks, providing some rough hints about how to proceed when I am less tired~~.  After reverting some overzealous requirement propagation, all proofs are working now.

### CODE REVIEW

File: src/LibraBFT/Impl/Consensus/BlockStorage/Properties/BlockStore.agda [hunk 1]
```diff
@@ -160,14 +160,19 @@ module executeAndInsertBlockESpec (bs0 : BlockStore) (vblock : ValidBlock) where
...
-             con⇒bindPost (Right (bt' , eb')) con' ._ refl = mkContractOk IBE.blocks≈ btP bss≡x
+             con⇒bindPost (Right (bt' , eb')) con' ._ refl =
+               mkContractOk (λ nohc → IBE.blocks≈ nohc block-c)
+                            btP bss≡x
                where
                  module IBE = insertBlockESpec.ContractOk con'
 
                  open BlockStoreInv
 
+                 postulate -- TODO: placeholder, need to figure out how to get this here
```
Comment (mark-moir) [inline src/LibraBFT/Impl/Consensus/BlockStorage/Properties/BlockStore.agda:171]: Ooops, not quite as done as I thought :smile:.

**Review by haroldcarr — COMMENTED** [review]
Please split this into two pull requests:
1. Core, Either, Maybe
2. LBFT

File: src/LibraBFT/Impl/Consensus/BlockStorage/Properties/BlockStore.agda (hunk 1 above)
Comment (mark-moir) [reply src/LibraBFT/Impl/Consensus/BlockStorage/Properties/BlockStore.agda]: Done now

**Comment by mark-moir** [comment]
> Please split this into two pull requests:
> 
> 1. Core, Either, Maybe
> 2. LBFT

Good idea, will do, thanks.

**Pull request closed without merging**